src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java

```diff
@@ -125,6 +133,312 @@ public GermplasmServiceImpl(final HibernateSessionProvider sessionProvider) {
 		return results;
 	}
 
+	@Override
+	public Set<Integer> importGermplasmUpdates(final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+
+		final Multimap<String, Object[]> conflictErrors = ArrayListMultimap.create();
+
+		// Get the names as well as the codes specified in the preferred name property.
+		final Set<String> namesCode = new HashSet<>();
+		germplasmUpdateDTOList.forEach(
+			g -> namesCode.addAll(g.getNames().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		namesCode
+			.addAll(germplasmUpdateDTOList.stream().map(o -> o.getPreferredName()).filter(Objects::nonNull).collect(Collectors.toSet()));
+
+		final Set<String> attributesCode = new HashSet<>();
+		germplasmUpdateDTOList.forEach(
+			g -> attributesCode.addAll(g.getAttributes().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+
+		// Retrieve the field id of attributes and names
+		final Map<String, Integer> attributeCodesFieldNoMap =
+			this.daoFactory.getUserDefinedFieldDAO().getByCodes(UDTableType.ATRIBUTS_ATTRIBUTE.getTable(),
+				new HashSet<>(Arrays.asList(UDTableType.ATRIBUTS_ATTRIBUTE.getType(), UDTableType.ATRIBUTS_PASSPORT.getType())),
+				attributesCode).stream().collect(Collectors.toMap(
+				UserDefinedField::getFcode, UserDefinedField::getFldno));
+		final Map<String, Integer> nameCodesFieldNoMap =
+			this.daoFactory.getUserDefinedFieldDAO()
+				.getByCodes(UDTableType.NAMES_NAME.getTable(),
+					Collections.singleton(UDTableType.NAMES_NAME.getType()), namesCode).stream().collect(Collectors.toMap(
+				UserDefinedField::getFcode, UserDefinedField::getFldno));
+
+		// germplasm UUID should be the priority in getting germplasm
+		final Set<String> germplasmUUIDs =
+			germplasmUpdateDTOList.stream().map(o -> o.getGermplasmUUID()).collect(Collectors.toSet());
+		// If there's no UUID, use GID
+		final Set<Integer> gids =
+			germplasmUpdateDTOList.stream().map(o -> liquibase.util.StringUtils.isEmpty(o.getGermplasmUUID()) ? o.getGid() : null).filter(
+				Objects::nonNull)
+				.collect(Collectors.toSet());
+
+		final List<Germplasm> germplasmList = new ArrayList<>();
```

Comment:
Same here the building of the germplasm list from specified gids and GUUIDs can be extracted to separate method:
`// germplasm UUID should be the priority in getting germplasm
		final Set<String> germplasmUUIDs =
			germplasmUpdateDTOList.stream().map(o -> o.getGermplasmUUID()).collect(Collectors.toSet());
		// If there's no UUID, use GID
		final Set<Integer> gids =
			germplasmUpdateDTOList.stream().map(o -> liquibase.util.StringUtils.isEmpty(o.getGermplasmUUID()) ? o.getGid() : null).filter(
				Objects::nonNull)
				.collect(Collectors.toSet());

		final List<Germplasm> germplasmList = new ArrayList<>();
		germplasmList.addAll(this.daoFactory.getGermplasmDao().getByUUIDListWithMethodAndBibref(germplasmUUIDs));
		germplasmList.addAll(this.daoFactory.getGermplasmDao().getByGIDListWithMethodAndBibref(gids));`